File: 11/src/GraphAM.cs
public class GraphAM : IGraph
{
    private Edge[,] adjacenz_matrix;
    public int NodeCount { get; private set; }
    public int EdgeCount { get; private set; }
    public bool IsDirected { get; private set; }
    public bool IsUndirected { get { return !IsDirected; } }

    public GraphAM(int node_count, bool is_directed)
    {
        NodeCount = node_count;
        IsDirected = is_directed;
        Clear();
    }

    public void AddEdge(int u, int v, double weight)
    {
        if (HasEdge(u,v))
            return;

        adjacenz_matrix[u-1, v-1] = new Edge(IsDirected, u, v, weight);

        if (IsUndirected)
            adjacenz_matrix[v-1, u-1] = new Edge(IsDirected, v, u, weight);;

        EdgeCount++;
    }

    public void DeleteEdge(int u, int v)
    {
        if (!HasEdge(u, v))
            return;

        adjacenz_matrix[u-1, v-1] = null;

        if (IsUndirected)
            adjacenz_matrix[v-1, u-1] = null;

        EdgeCount--;
    }

    public bool HasEdge(int u, int v)
    {
        return adjacenz_matrix[u-1, v-1] != null;
    }

    public IEnumerable<int> GetNeighborsOf(int u)
    {
        for (int v=1; v<=NodeCount; v++)
            if (HasEdge(u, v))
                yield return v;
    }

    public IEnumerable<Edge> GetEdgesFrom(int u)
    {
        for (int v=1; v<=NodeCount; v++)
            if (HasEdge(u, v))
                yield return adjacenz_matrix[u-1, v-1];
    }

    public double GetWeight(int u, int v)
    {
        if (HasEdge(u, v))
            return adjacenz_matrix[u-1, v-1].Weight;
        else
            return 0;
    }

    public IEnumerable<int> AllNodes
    {
        get
        {
            for (int i=1; i<=NodeCount; i++)
                yield return i;
        }
    }

    public void Clear()
    {
        adjacenz_matrix = new Edge[NodeCount, NodeCount];
        EdgeCount = 0;
    }

    public IEnumerable<Edge> AllEdges
    {
        get 
        {
            var hs = new HashSet<Edge>();
            foreach (var node in AllNodes)
                foreach (var edge in GetEdgesFrom(node))
                   hs.Add(edge);

            return hs;
        }
    }
}

File: hshl/AlgorithmsInCpp/src/GraphAM.cpp
#include "GraphAM.h"
#include <stdio.h>

GraphAM::GraphAM(int count) {
    node_count = count;
    this->adjacency_matrix = new int[count*count];
    for (int i=0; i<count*count; i++)
        this->adjacency_matrix[i] = 0;
}

int GraphAM::nodeCount() {
    return node_count;
}

void GraphAM::addEdge(int u, int v) {
    adjacency_matrix[(u-1)*node_count+v-1] = 1;
}

void GraphAM::deleteEdge(int u, int v) {
    adjacency_matrix[(u-1)*node_count+v-1] = 0;
}

bool GraphAM::hasEdge(int u, int v) {
    return this->adjacency_matrix[(u-1)*node_count+v-1] == 1;
}

void GraphAM::print() {
    for (int i=0; i<node_count; i++) {
        for (int j=0; j<node_count; j++) {
            printf("%i ", this->adjacency_matrix[i*node_count+j]);
        }
        printf("\n");
    }
}

std::list<int> GraphAM::nodesFrom(int node) {
    std::list<int> nodes;
    for (int i=1; i<=node_count; i++)
        if (hasEdge(node, i))
            nodes.push_back(i);
    
    return nodes;
}
